private/buf/buffetch/source_ref.go

```diff
@@ -24,22 +24,55 @@ import (
 var _ SourceRef = &sourceRef{}
 
 type sourceRef struct {
-	bucketRef internal.BucketRef
-	dirPath   string
+	bucketRef  internal.BucketRef
+	dirPath    string
+	subDirPath string
 }
 
 func newSourceRef(bucketRef internal.BucketRef) *sourceRef {
 	var dirPath string
 	if dirRef, ok := bucketRef.(internal.DirRef); ok {
 		dirPath = dirRef.Path()
 	}
+	var subDirPath string
+	if archiveRef, ok := bucketRef.(internal.ArchiveRef); ok {
+		subDirPath = archiveRef.SubDirPath()
+	}
+	if subDirPath == "" {
+		// Note that a bucketRef cannot be both an ArchiveRef and
+		// a GitRef, but we encapsulate this in an empty check just
+		// to be safe.
+		if gitRef, ok := bucketRef.(internal.GitRef); ok {
+			subDirPath = gitRef.SubDirPath()
+		}
+	}
 	return &sourceRef{
-		bucketRef: bucketRef,
-		dirPath:   dirPath,
+		bucketRef:  bucketRef,
+		dirPath:    dirPath,
+		subDirPath: subDirPath,
 	}
 }
 
 func (r *sourceRef) PathForExternalPath(externalPath string) (string, error) {
+	if r.subDirPath != "" {
+		path, err := normalpath.NormalizeAndValidate(externalPath)
```

Comment:
what if externalPath is absolute? is it that if subDirPath is set, this will never be true? this all needs more explanation (including in the code)